File: android/Airbridge/app/src/main/java/com/airbridge/mirror/MirrorMessage.kt
package com.airbridge.mirror

import java.nio.ByteBuffer
import java.nio.ByteOrder

enum class MirrorStatusCode(val raw: Byte) {
    SCREEN_OFF(1),
    APP_BACKGROUNDED(2),
    ACCESSIBILITY_DISABLED(3),
    ENCODER_ERROR(4),
    ACCESSIBILITY_BLOCKED(5);

    companion object {
        fun fromRaw(raw: Byte): MirrorStatusCode? = values().firstOrNull { it.raw == raw }
    }
}

class MirrorMessageException(message: String) : RuntimeException(message)

@OptIn(ExperimentalUnsignedTypes::class)
sealed class MirrorMessage {
    data class Hello(
        val token: ByteArray,
        val screenWidth: UInt,
        val screenHeight: UInt,
        val orientation: UByte
    ) : MirrorMessage() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Hello) return false
            return token.contentEquals(other.token) &&
                screenWidth == other.screenWidth &&
                screenHeight == other.screenHeight &&
                orientation == other.orientation
        }
        override fun hashCode(): Int =
            (((token.contentHashCode() * 31) + screenWidth.hashCode()) * 31 + screenHeight.hashCode()) * 31 + orientation.hashCode()
    }

    data class HelloAck(
        val targetBitrateBps: UInt,
        val fps: UByte,
        val keyframeIntervalSeconds: UByte,
        val targetWidth: UInt,
        val targetHeight: UInt,
        val codec: UByte  // 0 = H.264, 1 = HEVC
    ) : MirrorMessage()

    data class VideoConfig(val sps: ByteArray, val pps: ByteArray) : MirrorMessage() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is VideoConfig) return false
            return sps.contentEquals(other.sps) && pps.contentEquals(other.pps)
        }
        override fun hashCode() = sps.contentHashCode() * 31 + pps.contentHashCode()
    }

    data class VideoConfigHEVC(val vps: ByteArray, val sps: ByteArray, val pps: ByteArray) : MirrorMessage() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is VideoConfigHEVC) return false
            return vps.contentEquals(other.vps) && sps.contentEquals(other.sps) && pps.contentEquals(other.pps)
        }
        override fun hashCode() = (vps.contentHashCode() * 31 + sps.contentHashCode()) * 31 + pps.contentHashCode()
    }

    data class VideoFrame(val presentationTimestampUs: ULong, val naluBytes: ByteArray) : MirrorMessage() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is VideoFrame) return false
            return presentationTimestampUs == other.presentationTimestampUs && naluBytes.contentEquals(other.naluBytes)
        }
        override fun hashCode() = presentationTimestampUs.hashCode() * 31 + naluBytes.contentHashCode()
    }

    data class InputTap(
        val xNorm: Float,
        val yNorm: Float
    ) : MirrorMessage()

    data class Status(val code: MirrorStatusCode) : MirrorMessage()

    /** Reverse mirror: phone -> Mac, "start sending me YOUR screen". Carries the
     *  phone's screen size and mode (0 = mirror Mac main display, 1 = virtual
     *  display shaped to the phone). The Mac then replies with VideoConfig/
     *  VideoFrame down this same connection. */
    data class ReverseHello(
        val token: ByteArray,
        val screenWidth: UInt,
        val screenHeight: UInt,
        val mode: UByte
    ) : MirrorMessage() {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is ReverseHello) return false
            return token.contentEquals(other.token) &&
                screenWidth == other.screenWidth &&
                screenHeight == other.screenHeight &&
                mode == other.mode
        }
        override fun hashCode() =
            (((token.contentHashCode() * 31) + screenWidth.hashCode()) * 31 + screenHeight.hashCode()) * 31 + mode.hashCode()
    }

    /** Reverse control: phone -> Mac pointer. type: 0=move,1=down,2=up,3=drag. */
    data class ReverseInput(val type: UByte, val xNorm: Float, val yNorm: Float) : MirrorMessage()

    /** Reverse control: phone -> Mac scroll wheel (points). */
    data class ReverseScroll(val deltaX: Float, val deltaY: Float) : MirrorMessage()

    companion object {
        private const val TYPE_HELLO: Byte = 0x01
        private const val TYPE_HELLO_ACK: Byte = 0x02
        private const val TYPE_VIDEO_CONFIG: Byte = 0x10
        private const val TYPE_VIDEO_FRAME: Byte = 0x11
        private const val TYPE_VIDEO_CONFIG_HEVC: Byte = 0x12
        private const val TYPE_INPUT_TAP: Byte = 0x20
        private const val TYPE_STATUS: Byte = 0x30
        private const val TYPE_REVERSE_HELLO: Byte = 0x40
        private const val TYPE_REVERSE_INPUT: Byte = 0x41
        private const val TYPE_REVERSE_SCROLL: Byte = 0x42

        fun decode(bytes: ByteArray): MirrorMessage {
            if (bytes.isEmpty()) throw MirrorMessageException("empty payload")
            val buf = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)
            return when (val type = buf.get()) {
                TYPE_HELLO -> {
                    if (bytes.size != 1 + 16 + 4 + 4 + 1) throw MirrorMessageException("HELLO truncated")
                    val token = ByteArray(16); buf.get(token)
                    val w = buf.int.toUInt(); val h = buf.int.toUInt()
                    val orient = buf.get().toUByte()
                    Hello(token, w, h, orient)
                }
                TYPE_HELLO_ACK -> {
                    if (bytes.size < 1 + 4 + 1 + 1 + 4 + 4) throw MirrorMessageException("HELLO_ACK truncated")
                    val bitrate = buf.int.toUInt()
                    val fps = buf.get().toUByte()
                    val kf = buf.get().toUByte()
                    val w = buf.int.toUInt(); val h = buf.int.toUInt()
                    val codec = if (buf.hasRemaining()) buf.get().toUByte() else 0u.toUByte()  // absent = H.264
                    HelloAck(bitrate, fps, kf, w, h, codec)
                }
                TYPE_VIDEO_CONFIG -> {
                    if (buf.remaining() < 4) throw MirrorMessageException("VIDEO_CONFIG short SPS len")
                    val spsLen = buf.int
                    if (buf.remaining() < spsLen + 4) throw MirrorMessageException("VIDEO_CONFIG short SPS")
                    val sps = ByteArray(spsLen); buf.get(sps)
                    val ppsLen = buf.int
                    if (buf.remaining() != ppsLen) throw MirrorMessageException("VIDEO_CONFIG bad PPS")
                    val pps = ByteArray(ppsLen); buf.get(pps)
                    VideoConfig(sps, pps)
                }
                TYPE_VIDEO_CONFIG_HEVC -> {
                    if (buf.remaining() < 4) throw MirrorMessageException("VIDEO_CONFIG_HEVC short VPS len")
                    val vpsLen = buf.int
                    if (buf.remaining() < vpsLen + 4) throw MirrorMessageException("VIDEO_CONFIG_HEVC short VPS")
                    val vps = ByteArray(vpsLen); buf.get(vps)
                    val spsLen = buf.int
                    if (buf.remaining() < spsLen + 4) throw MirrorMessageException("VIDEO_CONFIG_HEVC short SPS")
                    val sps = ByteArray(spsLen); buf.get(sps)
                    val ppsLen = buf.int
                    if (buf.remaining() != ppsLen) throw MirrorMessageException("VIDEO_CONFIG_HEVC bad PPS")
                    val pps = ByteArray(ppsLen); buf.get(pps)
                    VideoConfigHEVC(vps, sps, pps)
                }
                TYPE_VIDEO_FRAME -> {
                    if (buf.remaining() < 8) throw MirrorMessageException("VIDEO_FRAME no PTS")
                    val pts = buf.long.toULong()
                    val nalu = ByteArray(buf.remaining()); buf.get(nalu)
                    VideoFrame(pts, nalu)
                }
                TYPE_INPUT_TAP -> {
                    if (buf.remaining() != 8) throw MirrorMessageException("INPUT_TAP truncated")
                    InputTap(buf.float, buf.float)
                }
                TYPE_STATUS -> {
                    if (buf.remaining() != 1) throw MirrorMessageException("STATUS truncated")
                    val code = MirrorStatusCode.fromRaw(buf.get())
                        ?: throw MirrorMessageException("STATUS unknown code")
                    Status(code)
                }
                TYPE_REVERSE_HELLO -> {
                    if (bytes.size != 1 + 16 + 4 + 4 + 1) throw MirrorMessageException("REVERSE_HELLO truncated")
                    val token = ByteArray(16); buf.get(token)
                    val w = buf.int.toUInt(); val h = buf.int.toUInt()
                    val mode = buf.get().toUByte()
                    ReverseHello(token, w, h, mode)
                }
                TYPE_REVERSE_INPUT -> {
                    if (buf.remaining() != 1 + 4 + 4) throw MirrorMessageException("REVERSE_INPUT truncated")
                    ReverseInput(buf.get().toUByte(), buf.float, buf.float)
                }
                TYPE_REVERSE_SCROLL -> {
                    if (buf.remaining() != 4 + 4) throw MirrorMessageException("REVERSE_SCROLL truncated")
                    ReverseScroll(buf.float, buf.float)
                }
                else -> throw MirrorMessageException("unknown type 0x${type.toUByte().toString(16)}")
            }
        }
    }

    fun encode(): ByteArray = when (this) {
        is Hello -> ByteBuffer.allocate(1 + 16 + 4 + 4 + 1).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_HELLO).put(token).putInt(screenWidth.toInt()).putInt(screenHeight.toInt()).put(orientation.toByte()).array()
        is HelloAck -> ByteBuffer.allocate(1 + 4 + 1 + 1 + 4 + 4 + 1).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_HELLO_ACK).putInt(targetBitrateBps.toInt()).put(fps.toByte()).put(keyframeIntervalSeconds.toByte())
            .putInt(targetWidth.toInt()).putInt(targetHeight.toInt()).put(codec.toByte()).array()
        is VideoConfig -> ByteBuffer.allocate(1 + 4 + sps.size + 4 + pps.size).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_VIDEO_CONFIG).putInt(sps.size).put(sps).putInt(pps.size).put(pps).array()
        is VideoConfigHEVC -> ByteBuffer.allocate(1 + 4 + vps.size + 4 + sps.size + 4 + pps.size).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_VIDEO_CONFIG_HEVC).putInt(vps.size).put(vps).putInt(sps.size).put(sps).putInt(pps.size).put(pps).array()
        is VideoFrame -> ByteBuffer.allocate(1 + 8 + naluBytes.size).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_VIDEO_FRAME).putLong(presentationTimestampUs.toLong()).put(naluBytes).array()
        is InputTap -> ByteBuffer.allocate(1 + 4 + 4).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_INPUT_TAP).putFloat(xNorm).putFloat(yNorm).array()
        is Status -> byteArrayOf(TYPE_STATUS, code.raw)
        is ReverseHello -> ByteBuffer.allocate(1 + 16 + 4 + 4 + 1).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_REVERSE_HELLO).put(token)
            .putInt(screenWidth.toInt()).putInt(screenHeight.toInt()).put(mode.toByte()).array()
        is ReverseInput -> ByteBuffer.allocate(1 + 1 + 4 + 4).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_REVERSE_INPUT).put(type.toByte()).putFloat(xNorm).putFloat(yNorm).array()
        is ReverseScroll -> ByteBuffer.allocate(1 + 4 + 4).order(ByteOrder.BIG_ENDIAN)
            .put(TYPE_REVERSE_SCROLL).putFloat(deltaX).putFloat(deltaY).array()
    }
}

File: macos/Airbridge/Sources/Mirror/MirrorMessage.swift
import Foundation

public enum MirrorStatusCode: UInt8, CaseIterable, Sendable {
    case screenOff = 1
    case appBackgrounded = 2
    case accessibilityDisabled = 3
    case encoderError = 4
    case accessibilityBlocked = 5
}

public enum MirrorMessageError: Error, Equatable, Sendable {
    case empty
    case unknownType(UInt8)
    case truncated(type: UInt8)
}

public enum MirrorMessage: Equatable, Sendable {
    case hello(token: Data, screenWidth: UInt32, screenHeight: UInt32, orientation: UInt8)
    /// codec: 0 = H.264, 1 = HEVC. Tells the (encoding) phone which codec to use.
    case helloAck(targetBitrateBps: UInt32, fps: UInt8, keyframeIntervalSeconds: UInt8, targetWidth: UInt32, targetHeight: UInt32, codec: UInt8)
    case videoConfig(sps: Data, pps: Data)
    /// HEVC parameter sets (VPS, SPS, PPS). Self-describes the stream as HEVC.
    case videoConfigHEVC(vps: Data, sps: Data, pps: Data)
    case videoFrame(presentationTimestampUs: UInt64, naluBytes: Data)
    case inputTap(xNorm: Float32, yNorm: Float32)
    case status(MirrorStatusCode)
    /// Reverse mirror: phone -> Mac, "start sending me YOUR screen". Carries the
    /// phone's screen size and the mode (0 = mirror Mac's main display,
    /// 1 = create a virtual display shaped to the phone). The Mac then sends
    /// videoConfig/videoFrame down this connection (Mac -> phone).
    case reverseHello(token: Data, screenWidth: UInt32, screenHeight: UInt32, mode: UInt8)
    /// Reverse control: phone -> Mac pointer input on the captured display.
    /// type: 0 = move, 1 = down, 2 = up, 3 = drag. Coords normalized 0..1.
    case reverseInput(type: UInt8, xNorm: Float32, yNorm: Float32)
    /// Reverse control: phone -> Mac scroll wheel (points).
    case reverseScroll(deltaX: Float32, deltaY: Float32)

    private enum TypeByte {
        static let hello: UInt8 = 0x01
        static let helloAck: UInt8 = 0x02
        static let videoConfig: UInt8 = 0x10
        static let videoFrame: UInt8 = 0x11
        static let videoConfigHEVC: UInt8 = 0x12
        static let inputTap: UInt8 = 0x20
        static let status: UInt8 = 0x30
        static let reverseHello: UInt8 = 0x40
        static let reverseInput: UInt8 = 0x41
        static let reverseScroll: UInt8 = 0x42
    }

    public func encode() -> Data {
        var out = Data()
        switch self {
        case let .hello(token, w, h, orientation):
            out.append(TypeByte.hello)
            out.append(token)
            out.appendBE(w)
            out.appendBE(h)
            out.append(orientation)
        case let .helloAck(bitrate, fps, keyframe, w, h, codec):
            out.append(TypeByte.helloAck)
            out.appendBE(bitrate)
            out.append(fps)
            out.append(keyframe)
            out.appendBE(w)
            out.appendBE(h)
            out.append(codec)
        case let .videoConfig(sps, pps):
            out.append(TypeByte.videoConfig)
            out.appendBE(UInt32(sps.count))
            out.append(sps)
            out.appendBE(UInt32(pps.count))
            out.append(pps)
        case let .videoConfigHEVC(vps, sps, pps):
            out.append(TypeByte.videoConfigHEVC)
            out.appendBE(UInt32(vps.count))
            out.append(vps)
            out.appendBE(UInt32(sps.count))
            out.append(sps)
            out.appendBE(UInt32(pps.count))
            out.append(pps)
        case let .videoFrame(pts, nalu):
            out.append(TypeByte.videoFrame)
            out.appendBE(pts)
            out.append(nalu)
        case let .inputTap(xNorm, yNorm):
            out.append(TypeByte.inputTap)
            out.appendBE(xNorm.bitPattern)
            out.appendBE(yNorm.bitPattern)
        case let .status(code):
            out.append(TypeByte.status)
            out.append(code.rawValue)
        case let .reverseHello(token, w, h, mode):
            out.append(TypeByte.reverseHello)
            out.append(token)
            out.appendBE(w)
            out.appendBE(h)
            out.append(mode)
        case let .reverseInput(type, x, y):
            out.append(TypeByte.reverseInput)
            out.append(type)
            out.appendBE(x.bitPattern)
            out.appendBE(y.bitPattern)
        case let .reverseScroll(dx, dy):
            out.append(TypeByte.reverseScroll)
            out.appendBE(dx.bitPattern)
            out.appendBE(dy.bitPattern)
        }
        return out
    }

    public static func decode(_ data: Data) throws -> MirrorMessage {
        guard let first = data.first else { throw MirrorMessageError.empty }
        let payload = data.dropFirst()
        switch first {
        case TypeByte.hello:
            guard payload.count == 16 + 4 + 4 + 1 else { throw MirrorMessageError.truncated(type: first) }
            let token = Data(payload.prefix(16))
            var i = payload.startIndex + 16
            let w: UInt32 = payload.readBE(at: &i)
            let h: UInt32 = payload.readBE(at: &i)
            let orientation = payload[i]
            return .hello(token: token, screenWidth: w, screenHeight: h, orientation: orientation)

        case TypeByte.helloAck:
            guard payload.count >= 4 + 1 + 1 + 4 + 4 else { throw MirrorMessageError.truncated(type: first) }
            var i = payload.startIndex
            let bitrate: UInt32 = payload.readBE(at: &i)
            let fps = payload[i]; i += 1
            let keyframe = payload[i]; i += 1
            let w: UInt32 = payload.readBE(at: &i)
            let h: UInt32 = payload.readBE(at: &i)
            let codec: UInt8 = i < payload.endIndex ? payload[i] : 0   // back-compat: absent = H.264
            return .helloAck(targetBitrateBps: bitrate, fps: fps, keyframeIntervalSeconds: keyframe, targetWidth: w, targetHeight: h, codec: codec)

        case TypeByte.videoConfig:
            guard payload.count >= 8 else { throw MirrorMessageError.truncated(type: first) }
            var i = payload.startIndex
            let spsLen: UInt32 = payload.readBE(at: &i)
            guard payload.count >= 4 + Int(spsLen) + 4 else { throw MirrorMessageError.truncated(type: first) }
            let sps = Data(payload[i..<i + Int(spsLen)]); i += Int(spsLen)
            let ppsLen: UInt32 = payload.readBE(at: &i)
            guard payload.count == 4 + Int(spsLen) + 4 + Int(ppsLen) else { throw MirrorMessageError.truncated(type: first) }
            let pps = Data(payload[i..<i + Int(ppsLen)])
            return .videoConfig(sps: sps, pps: pps)

        case TypeByte.videoConfigHEVC:
            guard payload.count >= 12 else { throw MirrorMessageError.truncated(type: first) }
            var i = payload.startIndex
            let vpsLen: UInt32 = payload.readBE(at: &i)
            guard payload.count >= 4 + Int(vpsLen) + 8 else { throw MirrorMessageError.truncated(type: first) }
            let vps = Data(payload[i..<i + Int(vpsLen)]); i += Int(vpsLen)
            let spsLen: UInt32 = payload.readBE(at: &i)
            guard payload.count >= 4 + Int(vpsLen) + 4 + Int(spsLen) + 4 else { throw MirrorMessageError.truncated(type: first) }
            let sps = Data(payload[i..<i + Int(spsLen)]); i += Int(spsLen)
            let ppsLen: UInt32 = payload.readBE(at: &i)
            guard payload.count == 4 + Int(vpsLen) + 4 + Int(spsLen) + 4 + Int(ppsLen) else { throw MirrorMessageError.truncated(type: first) }
            let pps = Data(payload[i..<i + Int(ppsLen)])
            return .videoConfigHEVC(vps: vps, sps: sps, pps: pps)

        case TypeByte.videoFrame:
            guard payload.count >= 8 else { throw MirrorMessageError.truncated(type: first) }
            var i = payload.startIndex
            let pts: UInt64 = payload.readBE(at: &i)
            let nalu = Data(payload[i...])
            return .videoFrame(presentationTimestampUs: pts, naluBytes: nalu)

        case TypeByte.inputTap:
            guard payload.count == 8 else { throw MirrorMessageError.truncated(type: first) }
            var i = payload.startIndex
            let xBits: UInt32 = payload.readBE(at: &i)
            let yBits: UInt32 = payload.readBE(at: &i)
            return .inputTap(xNorm: Float32(bitPattern: xBits), yNorm: Float32(bitPattern: yBits))

        case TypeByte.status:
            guard payload.count == 1, let code = MirrorStatusCode(rawValue: payload[payload.startIndex]) else {
                throw MirrorMessageError.truncated(type: first)
            }
            return .status(code)

        case TypeByte.reverseHello:
            guard payload.count == 16 + 4 + 4 + 1 else { throw MirrorMessageError.truncated(type: first) }
            let token = Data(payload.prefix(16))
            var i = payload.startIndex + 16
            let w: UInt32 = payload.readBE(at: &i)
            let h: UInt32 = payload.readBE(at: &i)
            let mode = payload[i]
            return .reverseHello(token: token, screenWidth: w, screenHeight: h, mode: mode)

        case TypeByte.reverseInput:
            guard payload.count == 1 + 4 + 4 else { throw MirrorMessageError.truncated(type: first) }
            var i = payload.startIndex
            let type = payload[i]; i += 1
            let xBits: UInt32 = payload.readBE(at: &i)
            let yBits: UInt32 = payload.readBE(at: &i)
            return .reverseInput(type: type, xNorm: Float32(bitPattern: xBits), yNorm: Float32(bitPattern: yBits))

        case TypeByte.reverseScroll:
            guard payload.count == 4 + 4 else { throw MirrorMessageError.truncated(type: first) }
            var i = payload.startIndex
            let dxBits: UInt32 = payload.readBE(at: &i)
            let dyBits: UInt32 = payload.readBE(at: &i)
            return .reverseScroll(deltaX: Float32(bitPattern: dxBits), deltaY: Float32(bitPattern: dyBits))

        default:
            throw MirrorMessageError.unknownType(first)
        }
    }
}

// MARK: - Big-endian helpers

private extension Data {
    mutating func appendBE(_ value: UInt32) {
        var be = value.bigEndian
        Swift.withUnsafeBytes(of: &be) { append(contentsOf: $0) }
    }
    mutating func appendBE(_ value: UInt64) {
        var be = value.bigEndian
        Swift.withUnsafeBytes(of: &be) { append(contentsOf: $0) }
    }
    func readBE(at i: inout Int) -> UInt32 {
        let v = self[i..<i+4].withUnsafeBytes { $0.loadUnaligned(as: UInt32.self) }.bigEndian
        i += 4
        return v
    }
    func readBE(at i: inout Int) -> UInt64 {
        let v = self[i..<i+8].withUnsafeBytes { $0.loadUnaligned(as: UInt64.self) }.bigEndian
        i += 8
        return v
    }
}
